flasksrc/session.py

```diff
@@ -607,6 +625,105 @@ def put(self, toynet_session_id, **kwargs):
             }, 200
 
 
+class ToyNetSessionByIdCreateRouterInterface(MethodResource):
+    @use_kwargs(ToyNetSessionByIdCreateRouterInterfacePutReq)
+    @marshal_with(ToyNetSessionByIdCreateRouterInterfacePutResp)
+    def put(self, toynet_session_id, **kwargs):
+        try:
+            req = ToyNetSessionByIdCreateRouterInterfacePutReq().load(kwargs)
+        except ValidationError as e:
+            abort(400, message='Invalid request: {}'.format(e))
+
+        # Separate validation from Marshmallow
+        if 'ip' not in req:
+            abort(400, message='Missing ip from req')
+        elif 'name' not in req:
+            abort(400, message='Missing name from req')
+
+        orig_topology = getTopologyFromDb(toynet_session_id)['topology']
+        root = ET.fromstring(orig_topology)
+
+        # Get the router to add the interface to, note that we do not validate
+        # that the IP address is "good" (eg: not a duplicate of another
+        # interface)
+        router = None
+        router_list = root.find('routerList').findall('router')
+
+        for device in router_list:
+            if device.get('name') == req['name']:
+                router = device
+
+        if router is None:
+            abort(400, message=f'Router {req["name"]} does not exist')
+
+        # Create XML elements and update topology
+        interface = ET.Element('intf')
+        interface.text = req['ip']
+        router.append(interface)
+        new_topology = ET.tostring(root).decode('utf-8')
+
+        # Send to mininet and update DB, these functions abort the REST call on
+        # failure
+        sendTopoToMininet(toynet_session_id, new_topology)
+        updateTopoInDb(toynet_session_id, new_topology)
+
+        return {
+            }, 200
+
+
+class ToyNetSessionByIdDeleteRouterInterface(MethodResource):
+    @use_kwargs(ToyNetSessionByIdDeleteRouterInterfacePutReq)
+    @marshal_with(ToyNetSessionByIdDeleteRouterInterfacePutResp)
+    def put(self, toynet_session_id, **kwargs):
+        try:
+            req = ToyNetSessionByIdDeleteRouterInterfacePutReq().load(kwargs)
+        except ValidationError as e:
+            abort(400, message='Invalid request: {}'.format(e))
+
+        # Separate validation from Marshmallow
+        if 'ip' not in req:
+            abort(400, message='Missing ip from req')
+        elif 'name' not in req:
+            abort(400, message='Missing name from req')
+
+        orig_topology = getTopologyFromDb(toynet_session_id)['topology']
+        root = ET.fromstring(orig_topology)
+
+        # Update the topology and get the interface index
+        intf_idx = None
+        for device in root.find('routerList'):
+            if device.attrib['name'] == req['name']:
+                for idx, intf in enumerate(device.findall('intf')):
```

Comment:
Although I haven't tried it out, you may be able to do a **list comprehension** somewhere in here. You might also be able to do a `next` instead of breaking.

* https://stackoverflow.com/questions/6981717/pythonic-way-to-combine-for-loop-and-if-statement
* https://stackoverflow.com/questions/8534256/find-first-element-in-a-sequence-that-matches-a-predicate